### ISSUE TITLE
CTC Loss for Speech Recognition models

### ISSUE THREAD

**Issue by patrickvonplaten** [google/jax#9350]
Please:

- [x] Check for duplicate requests.
- [x] Describe your goal, and if possible provide a code snippet with a motivating example.

## Community Request for JAX CTC Loss

First of all, I'm not 100% sure whether this repo is the right place or whether we should rather open it in https://github.com/deepmind/optax or https://github.com/google/flax

Many state-of-the-art speech recognition models are fine-tuned using the CTC loss - most notably Wav2Vec2. Both PyTorch and Tensorflow have native support for the CTC loss - see: 
- https://pytorch.org/docs/stable/generated/torch.nn.CTCLoss.html
- https://www.tensorflow.org/api_docs/python/tf/nn/ctc_loss

People seem to start using TPUs more often and PyTorch's CTC loss is not made for torch/XLA, see: https://github.com/pytorch/xla/issues/2681#issuecomment-894468034
JAX has very good TPU support and is arguably more user-friendly for PyTorch users than TF.

Think this could also be a cool first/second issue. 

Gently pinging @marcvanzee  @jheek @avital here as well

**Comment by jheek** [comment]
It could also live as an objective in [JaxOpt](https://jaxopt.github.io/stable/objective_and_loss.html)

**Comment by patrickvonplaten** [comment]
Cool, would do you guys prefer? Think we could find some community members to work on adding such a loss actually

**Comment by patrickvonplaten** [comment]
In case this is already somewhere developed internally and will soon be open-sourced it'd be great to hear about it :-)

**Comment by mattjj** [comment]
We're also asking the optax folks what they think. I'll assign this to myself so I remember to follow up when we hear things.

**Comment by patrickvonplaten** [comment]
Thanks @hawkinsp,

I'll try this one out for fine-tune a Wav2Vec2 model in Flax and report back here if results are similar / equal to PyTorch

**Comment by frmccann97** [comment]
Did you get good results?

**Comment by frmccann97** [comment]
> Thanks @hawkinsp,
> 
> I'll try this one out for fine-tune a Wav2Vec2 model in Flax and report back here if results are similar / equal to PyTorch

How did it go?

**Comment by redwrasse** [comment]
Following up on this thread. Is a ctc loss in jax of interest to the maintainers, or is a preference to rely on downstream libraries?

**Comment by yotarok** [comment]
Hi @redwrasse,

Optax has a CTC loss implementation.
https://github.com/google-deepmind/optax/blob/00e07e23104452e08bb46d79bb288413e622574d/optax/losses/_classification.py#L760

This is generic enough that you don't need to use the entire Optax framework, and you can use just this loss function.
